homeassistant/config_entries.py

```diff
@@ -1145,8 +1148,27 @@ def __setitem__(self, entry_id: str, entry: ConfigEntry) -> None:
         data[entry_id] = entry
         self._domain_index.setdefault(entry.domain, []).append(entry)
         if entry.unique_id is not None:
+            unique_id_hash = entry.unique_id
+            # Guard against integrations using unhashable unique_id
+            # In HA Core 2024.9, we should remove the guard and instead fail
+            if not isinstance(entry.unique_id, Hashable):
+                unique_id_hash = str(entry.unique_id)  # type: ignore[unreachable]
```

Comment:
I think we should not load the config entry if the data is incorrect. I don't think we should try to be smart about it.

Reply:
I don't agree

Reply:
By making it a string we are relying on the custom component author who created it with a non string to be able to clean it up later.

I guess we are stuck with an assumption that they aren't actually using the unique id to store a list of data that they need to set up the integration. I'm not sure if it's a reasonable position to accommodate that use case as it's pretty far away from how the API is expected to be used.

The safest course would be to revert the original PR and give custom component authors a few months to fix their integrations but unfortunately there is already other work built on top of it. So there is probably too much risk to do that.

For most cases it probably works fine to stringify it and load it.  It's the most user friendly option at least.

Reply:
The only semi-realistic risk I can come up with that seems possible is they have stored some object in the unique id that has a string representation that isn't unique but is somehow JSON serializable.  That would be some type of impressive abuse of the system so it seems like a very contrived case and acceptable risk.

Reply:
As example:
https://github.com/home-assistant/core/issues/109908#issuecomment-1934209009

Also, this is also still an issue for entity unique_ids (I found that out a few betas ago when I accidentally broke tellduslive)

Reply:
@joostlek I don't understand your comment, your example does not point to a case where there's a risk of duplicates

Reply:
I was responding to bdraco's comment about string representation and JSON serializable

Reply:
> I think we should not load the config entry if the data is incorrect. I don't think we should try to be smart about it.

We discussed this in a meeting and there were more in favor of allowing the config entries to load (as the PR does) than to not load the config entry (as Martin suggests).